docs/typescript-styleguide.md

```diff
@@ -0,0 +1,85 @@
+# Carbon TypeScript Style Guide
+
+## Follow these guidelines when contributing TypeScript to the Carbon repository
+
+### Contents
+
+[Introduction](#introduction)
+
+[Conventions](#conventions)
+
+[Linting](#linting)
+
+### Introduction
+
+This TypeScript style guide is provided to help you contribute to the Carbon repository in a way that other developers can understand quickly and easily.
+
+### Folder Structure
+
+We have a consistent approach to the file structure for our components as shown below:
+
+```
+src/components/new-component/
+  |-- __internal__/..
+  |-- sub-component/..
+  |-- new-component.js
+  |-- index.js
+  |-- new-component.style.js
+  |-- new-component.spec.js
+  |-- new-component.stories.mdx
+  |-- new-component.stories.js
+  |-- new-component.d.ts
+  |-- index.d.ts
+```
+
+Any sub-component that will not be exported directly from Carbon should go in the `new-component/__internal__/` folder. Sub-components should have the same file structure as shown for components where appropriate.
+
+Each component and sub-component should have a `<component-name>.d.ts` file containing it's TypeScript interface. The root directory of the component should also have a `index.d.ts` file which exports the interfaces from the component and any non-internal sub-components.
+
+### Conventions
+
+#### Component interface
+
+All component interfaces (in `<component-name>.d.ts` files) should follow this template:
+
+```ts
+import * as React from "react";
+
+export interface MyComponentProps {
+  /** Example required prop description */
+  exampleRequiredProp: string;
+  /** Example optional prop description */
+  exampleOptionalProp?: string;
+}
+
+declare const MyComponent: React.FunctionComponent<MyComponentProps>;
+
+export default MyComponent;
+```
+
+#### Index file
+
+The `index.d.ts` file for each component should follow this template:
+
+```ts
+export { default as MyComponent } from "./my-component";
+export { default as MySubComponent } from "./my-sub-component/my-sub-component";
```

Comment:
do we allow only named exports?

Reply:
As far as I've seen in Carbon, are you adding support for default exports in your current TypeScript work?

Reply:
I think that we can add also
```
export { default } from './my-component 
```
It's up to you, not sure it will help someone

Reply:
we're using default exports in:
Date, DateRange, Decimal, Fieldset, FormField...
I have counted 52 usages in our project

Reply:
Here's what the Date one looks like: https://github.com/Sage/carbon/blob/master/src/__experimental__/components/date/index.d.ts - which is exactly what I've got here

Are you looking in your branch where you're updating the ts interfaces?

Reply:
Ok I'm going to change it to this:

```ts
export { default } from "./my-component";
export { default  as MySubComponent } from "./my-sub-component/my-sub-component";
```

Then the main component is a default export, but the sub-components are not. The user can then import like: 

`import MyComponent, { MySubComponent } from 'carbon-react/lib/components/my-component';`